### DISCUSSION TITLE
Constant strain in cells - thermal strain simulation using Tensor mechanics

### DISCUSSION THREAD

**Discussion by HoWilgh** [idaholab/moose#26563] [Q&A Modules: Solid mechanics]
Hi, 

I have built a relatively simple model where I want to use the tensor mechanics module to simulate strain due to thermal expansion.
The model is attached and uses MOOSE-internally generated mesh.

Attached you find also screenshots from the evaluation of the results in Paraview.

One can see two main things for the given simulation:
* The displacement ist computed as point-data and the postprocessing using a line plot yields smooth (interpolated) values. See the upper two plots in the screenshot below.
* Strain and stress are computed as 'constant' volumetric (cell-) data and an evaluation over a line through the model yields steps in the strain-values.  See the lower two plots in the screenshot below.

How to overcome these 'constant'-computed vales for the strain to get as smooth results for the strain as for the displacement.

Thank you in advance,
HoWil
#############
Model:
```
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]


# Using internal mesh of MOOSE 
[Mesh]
    type = GeneratedMesh
    dim = 3
    nx = 10
    xmin = 0.0
    xmax = 200.0
    ny = 10
    ymin = 0.0
    ymax = 200.0
    nz = 10
    zmin = 0.0
    zmax = 100.0
    #displacements = 'disp_x disp_y disp_z'
    second_order = True
[]

[BCs]  # Roller BC for the outer surfaces
  [disp_x_right]
      type = DirichletBC
      variable = disp_x
      boundary = right
      value = 0.0
  []

  [disp_y_front]
      type = DirichletBC
      variable = disp_y
      boundary = front
      value = 0.0
  []

  [disp_z_bottom]
      type = DirichletBC
      variable = disp_z
      boundary = bottom
      value = 0.0
  []
[]

[Modules/TensorMechanics/Master]
  [all]
    add_variables = true
    automatic_eigenstrain_names = true
    strain = SMALL  # Small linearized strain, automatically set to XY coordinates
    generate_output = 'stress_xx strain_xx' #automatically creates the auxvariables and auxkernels
    #use_interpolated_state = true
    # incremental = true
  []
[]

[Materials]
  [elasticity1]
    type = ComputeElasticityTensor
    fill_method = symmetric9
    # C_ijkl = '1111  1122  1133  2222  2233  3333  2323  3131  1212'
    # C_ijkl = '11    12    13    22    23    33    44    55    66'
    C_ijkl = '234 92 88 234 88 238 64 64 54'
    rotation_matrix = '1.0 0.0 0.0
                      0.0 1.0 0.0
                      0.0 0.0 1.0'
  []

  # eigenstrain computation due to temperature
  [expansion1]
    type = ComputeEigenstrain
    eigen_base = '0.000375 0.0 0.0 0.0 0.000375 0.0 0.0 0.0 -0.0031'
    prefactor = 1.
    eigenstrain_name = eigenstrain
  []  

  [stress]
    type = ComputeLinearElasticStress
  []
  
[]

[Executioner]  # See https://mooseframework.inl.gov/modules/porous_flow/solvers.html
  type = Steady
  solve_type = LINEAR
[]

[Postprocessors]
  [nodes]
    type = NumNodes
  []
  [elements]
    type = NumElems
  []
[]

[Outputs]
  exodus = true
[]
```
#############
Screenshot of Paraview evaluating the computed results using a Plotoverline-filter:
![Screenshot from 2024-01-16 10-35-25](https://github.com/idaholab/moose/assets/9978035/d0298d9e-6bee-4652-b650-b38939c0011a)

**Comment by GiudGiud** [comment]
Hello

There are two options here.
- paraview can convert cell data to node data, and plot it linearly.
or you can use a ProjectionAux in MOOSE to project the output auxiliary variable onto a different FE family, like linear lagrange, that is smoother and still minimize the L2 norm between the two. 

Guillaume

**Reply by jiangwen84** [reply]
In addition to @GiudGiud 's suggestion, you can also use https://mooseframework.inl.gov/source/auxkernels/NodalPatchRecoveryAux.html
